File: agentic-loops/run_agent.py
#!/usr/bin/env python3
"""
Production Runtime Harness — Python
Wires react_loop.py into a fully runnable CLI agent.

Usage:
  python run_agent.py "your task here"
  python run_agent.py --system "You are a DevOps expert." "audit my /tmp directory"
  python run_agent.py --max-iter 50 --json-out /tmp/result.json "your task"
  echo "your task" | python run_agent.py -

Environment:
  ANTHROPIC_API_KEY   required
  AGENT_MAX_ITER      default 30
  AGENT_SYSTEM        default system prompt override
  AGENT_LOG_LEVEL     debug | info (default info)
"""

import argparse
import json
import logging
import os
import signal
import sys
import time
import traceback
from pathlib import Path

# ── Logging setup ─────────────────────────────────────────────────────────────
log_level = os.environ.get("AGENT_LOG_LEVEL", "info").upper()
logging.basicConfig(
    level=getattr(logging, log_level, logging.INFO),
    format="%(asctime)s [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%dT%H:%M:%S",
    stream=sys.stderr,
)
log = logging.getLogger("harness")

# ── Validate environment ──────────────────────────────────────────────────────
def validate_env():
    key = os.environ.get("ANTHROPIC_API_KEY", "").strip()
    if not key:
        log.error("ANTHROPIC_API_KEY is not set")
        sys.exit(1)
    if not key.startswith("sk-ant-"):
        log.warning("ANTHROPIC_API_KEY does not look like a valid Anthropic key")

# ── Signal handling ───────────────────────────────────────────────────────────
_interrupted = False

def _handle_sigint(sig, frame):
    global _interrupted
    if _interrupted:
        log.error("Force quit")
        sys.exit(130)
    _interrupted = True
    log.warning("Interrupt received — finishing current iteration then stopping...")

signal.signal(signal.SIGINT, _handle_sigint)

# ── Import engine ─────────────────────────────────────────────────────────────
try:
    from react_loop import run_react, Thought, Action, Observation, Answer
except ImportError as e:
    log.error("Cannot import react_loop.py — make sure it is in the same directory: %s", e)
    sys.exit(1)

# ── Step printer ──────────────────────────────────────────────────────────────
def make_step_printer(verbose: bool):
    def on_step(step):
        if _interrupted:
            raise KeyboardInterrupt("Agent interrupted by user")
        if isinstance(step, Thought):
            sys.stdout.write(f"\n\033[94m💭 THOUGHT\033[0m\n{step.content}\n")
        elif isinstance(step, Action):
            inp = json.dumps(step.input, ensure_ascii=False)
            sys.stdout.write(f"\n\033[93m⚡ ACTION\033[0m  {step.tool}\n{inp}\n")
        elif isinstance(step, Observation):
            color = "\033[91m" if step.is_error else "\033[92m"
            label = "❌ ERROR" if step.is_error else "👁 OBSERVE"
            preview = step.content[:600] + ("..." if len(step.content) > 600 else "")
            sys.stdout.write(f"\n{color}{label}\033[0m\n{preview}\n")
        sys.stdout.flush()
    return on_step if verbose else None

# ── Main ──────────────────────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser(description="Production ReAct Agent Harness")
    parser.add_argument("task", nargs="?", default="-",
                        help="Task string, or '-' to read from stdin")
    parser.add_argument("--system",    default=os.environ.get("AGENT_SYSTEM", ""),
                        help="System prompt addition")
    parser.add_argument("--max-iter",  type=int,
                        default=int(os.environ.get("AGENT_MAX_ITER", 30)),
                        help="Max iterations (default 30)")
    parser.add_argument("--json-out",  default=None,
                        help="Write full result JSON to this path")
    parser.add_argument("--quiet",     action="store_true",
                        help="Suppress step-by-step output")
    args = parser.parse_args()

    validate_env()

    # Read task
    if args.task == "-":
        task = sys.stdin.read().strip()
    else:
        task = args.task.strip()

    if not task:
        log.error("No task provided")
        sys.exit(1)

    log.info("Task: %s", task[:200])
    log.info("Max iterations: %d", args.max_iter)

    start = time.time()
    exit_code = 0

    try:
        answer, trace = run_react(
            user_message   = task,
            system_extra   = args.system,
            max_iterations = args.max_iter,
            on_tool_call   = make_step_printer(not args.quiet),
        )

        elapsed = round(time.time() - start, 2)

        sys.stdout.write(f"\n{'='*64}\n")
        sys.stdout.write(f"FINAL ANSWER\n{'='*64}\n")
        sys.stdout.write(answer + "\n")
        sys.stdout.write(f"{'='*64}\n")
        sys.stdout.write(f"Iterations: {trace.iterations} | Tokens: {trace.total_tokens} | Time: {elapsed}s\n")
        sys.stdout.flush()

        if args.json_out:
            out = {
                "task":         task,
                "answer":       answer,
                "iterations":   trace.iterations,
                "total_tokens": trace.total_tokens,
                "elapsed_s":    elapsed,
                "trace":        trace.to_dict(),
            }
            Path(args.json_out).write_text(json.dumps(out, indent=2, ensure_ascii=False))
            log.info("Result written to %s", args.json_out)

    except KeyboardInterrupt:
        log.warning("Agent stopped by user")
        exit_code = 130
    except RuntimeError as e:
        log.error("Agent failed: %s", e)
        exit_code = 1
    except Exception:
        log.error("Unexpected error:\n%s", traceback.format_exc())
        exit_code = 1

    sys.exit(exit_code)

if __name__ == "__main__":
    main()

File: agentic-loops/run_agent.rs
//! Production Runtime Harness — Rust
//! Wires react_loop.rs into a fully runnable CLI agent.
//!
//! Same Cargo.toml deps as react_loop.rs, plus:
//!   clap = { version = "4", features = ["derive"] }
//!
//! Usage:
//!   cargo run -- "your task here"
//!   cargo run -- --system "You are a DevOps expert." "audit /tmp"
//!   cargo run -- --max-iter 50 --json-out /tmp/result.json "task"
//!   echo "task" | cargo run -- -
//!
//! Environment:
//!   ANTHROPIC_API_KEY   required
//!   AGENT_MAX_ITER      default 30
//!   AGENT_SYSTEM        system prompt override

use std::env;
use std::fs;
use std::io::{self, Read};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Instant;

use anyhow::{anyhow, Result};
use clap::Parser;
use serde_json::json;

// Include engine (in a real project: mod react_loop; use react_loop::*)
// For single-file: compile both files together or use mod.
use crate::{run_react, ReactOptions, ToolRegistry, Step, Trace};

// ── CLI args ──────────────────────────────────────────────────────────────────
#[derive(Parser, Debug)]
#[command(name = "run_agent", about = "Production ReAct Agent Runtime Harness")]
struct Args {
    /// Task to execute, or '-' to read from stdin
    #[arg(default_value = "-")]
    task: String,

    /// Additional system prompt
    #[arg(long, default_value = "")]
    system: String,

    /// Maximum iterations
    #[arg(long, default_value_t = 30)]
    max_iter: usize,

    /// Write result JSON to this path
    #[arg(long)]
    json_out: Option<String>,

    /// Suppress step-by-step output
    #[arg(long)]
    quiet: bool,
}

// ── Validate environment ──────────────────────────────────────────────────────
fn validate_env() {
    let key = env::var("ANTHROPIC_API_KEY").unwrap_or_default();
    if key.trim().is_empty() {
        eprintln!("[harness] ERROR: ANTHROPIC_API_KEY is not set");
        std::process::exit(1);
    }
    if !key.trim_start().starts_with("sk-ant-") {
        eprintln!("[harness] WARN: ANTHROPIC_API_KEY does not look like a valid Anthropic key");
    }
}

// ── Step printer ──────────────────────────────────────────────────────────────
fn make_step_printer(verbose: bool, interrupted: Arc<AtomicBool>) -> Option<Box<dyn Fn(&Step) + Send>> {
    if !verbose { return None; }
    Some(Box::new(move |step: &Step| {
        if interrupted.load(Ordering::Relaxed) {
            eprintln!("[harness] Agent interrupted");
            std::process::exit(130);
        }
        match step {
            Step::Thought { content } => {
                println!("\n\x1b[94m💭 THOUGHT\x1b[0m\n{}", content);
            }
            Step::Action { tool, input } => {
                let inp = serde_json::to_string_pretty(input).unwrap_or_default();
                println!("\n\x1b[93m⚡ ACTION\x1b[0m  {}\n{}", tool, inp);
            }
            Step::Observation { content, error } => {
                let (color, label) = if *error {
                    ("\x1b[91m", "❌ ERROR")
                } else {
                    ("\x1b[92m", "👁 OBSERVE")
                };
                let preview: String = content.chars().take(600).collect();
                let suffix = if content.len() > 600 { "..." } else { "" };
                println!("\n{}{}\x1b[0m\n{}{}", color, label, preview, suffix);
            }
            _ => {}
        }
    }))
}

// ── Main ──────────────────────────────────────────────────────────────────────
#[tokio::main]
async fn main() -> Result<()> {
    validate_env();

    let mut args = Args::parse();

    // Override from env
    if let Ok(v) = env::var("AGENT_MAX_ITER") {
        if let Ok(n) = v.parse::<usize>() { args.max_iter = n; }
    }
    if let Ok(s) = env::var("AGENT_SYSTEM") {
        if args.system.is_empty() { args.system = s; }
    }

    // Read task
    let task = if args.task == "-" {
        let mut buf = String::new();
        io::stdin().read_to_string(&mut buf)?;
        buf.trim().to_string()
    } else {
        args.task.trim().to_string()
    };

    if task.is_empty() {
        eprintln!("[harness] ERROR: no task provided");
        std::process::exit(1);
    }

    eprintln!("[harness] Task: {}", &task[..task.len().min(200)]);
    eprintln!("[harness] Max iterations: {}", args.max_iter);

    // Signal handling
    let interrupted = Arc::new(AtomicBool::new(false));
    let interrupted_clone = interrupted.clone();
    ctrlc::set_handler(move || {
        if interrupted_clone.load(Ordering::Relaxed) {
            eprintln!("[harness] Force quit");
            std::process::exit(130);
        }
        interrupted_clone.store(true, Ordering::Relaxed);
        eprintln!("[harness] Interrupt — finishing current iteration...");
    }).ok();

    let registry = ToolRegistry::new();
    let start = Instant::now();

    let result = run_react(&task, &registry, ReactOptions {
        system_extra:   args.system.clone(),
        max_iterations: args.max_iter,
        on_step:        make_step_printer(!args.quiet, interrupted.clone()),
    }).await;

    let elapsed_ms = start.elapsed().as_millis();

    match result {
        Ok((answer, trace)) => {
            let sep = "=".repeat(64);
            println!("\n{}\nFINAL ANSWER\n{}\n{}\n{}", sep, sep, answer, sep);
            println!("Iterations: {} | Tokens: {} | Time: {}ms",
                trace.iterations, trace.total_tokens, elapsed_ms);

            if let Some(json_path) = &args.json_out {
                let doc = json!({
                    "task":         task,
                    "answer":       answer,
                    "iterations":   trace.iterations,
                    "total_tokens": trace.total_tokens,
                    "elapsed_ms":   elapsed_ms,
                    "steps":        trace.steps,
                });
                fs::write(json_path, serde_json::to_string_pretty(&doc)?)?;
                eprintln!("[harness] Result written to {}", json_path);
            }

            Ok(())
        }
        Err(e) => {
            if interrupted.load(Ordering::Relaxed) {
                eprintln!("[harness] Agent stopped by user");
                std::process::exit(130);
            }
            eprintln!("[harness] ERROR: {}", e);
            std::process::exit(1);
        }
    }
}
